File: projects/iris/src/rules/RuleEngine.ts
/**
 * RuleEngine — motore di regole declarative per IRIS.
 *
 * Una regola è un documento Firestore in `iris_rules`:
 * {
 *   id: string,
 *   name: string,
 *   description?: string,
 *   enabled: boolean,
 *   priority: number,         // più alto = valutato prima
 *   conditions: RuleCondition[], // ALL conditions must match (AND)
 *   actions: RuleAction[],
 *   stopOnMatch?: boolean,    // default true: ferma valutazione dopo match
 *   createdAt, updatedAt
 * }
 *
 * Strategia:
 *   1. loadRules() ordina per priority desc.
 *   2. evaluate(email) ritorna { matchedRule, actions } della prima regola
 *      che matcha (oppure null).
 *   3. execute(email, actions) esegue ogni azione tramite l'ActionRunner
 *      iniettato (dependency-injected per testabilità).
 *
 * Il motore NON conosce direttamente Firestore o la Lavagna: li riceve
 * come dipendenze. Questo lo rende testabile e riusabile (in pipeline
 * Python via porting, o in PWA per dry-run preview).
 */
import type { ClassificationType, SentimentLevel, SuggestedAction } from "../types/classification.js";
import type { LavagnaPriority } from "../../../nexo-core/lavagna/types.js";

// ─── Conditions ──────────────────────────────────────────────────────

export type ConditionField =
  | "sender"
  | "subject"
  | "category"
  | "sentiment"
  | "suggestedAction"
  | "hasAttachments"
  | "attachmentType"
  | "score";

export type ConditionOp =
  | "equals"
  | "not_equals"
  | "contains"
  | "not_contains"
  | "in"
  | "not_in"
  | "matches"     // regex
  | "gt"
  | "gte"
  | "lt"
  | "lte"
  | "is_true"
  | "is_false";

export interface RuleCondition {
  field: ConditionField;
  op: ConditionOp;
  value?: string | number | boolean | string[];
  /** case-insensitive per campi stringa (default true) */
  caseInsensitive?: boolean;
}

// ─── Actions ─────────────────────────────────────────────────────────

export type RuleActionType =
  | "write_lavagna"
  | "notify_echo"
  | "archive_email"
  | "set_priority"
  | "tag_email"
  | "extract_data";

export interface RuleAction {
  type: RuleActionType;
  /** target Lavagna (es. "ares", "charta", "dikea") per write_lavagna */
  to?: string;
  /** tipo messaggio Lavagna (es. "incassi_acg") per write_lavagna */
  messageType?: string;
  /** priorità Lavagna o priorità email per set_priority */
  priority?: LavagnaPriority;
  /** payload arbitrario per write_lavagna */
  payload?: Record<string, unknown>;
  /** canale ECHO ("wa" | "telegram" | "voice") */
  channel?: string;
  /** testo notifica per notify_echo */
  text?: string;
  /** tag list per tag_email */
  tags?: string[];
  /** patterns per extract_data, es. {"importo":"€\\s*([0-9.,]+)"} */
  extractPatterns?: Record<string, string>;
}

// ─── Rule ────────────────────────────────────────────────────────────

export interface IrisRule {
  id: string;
  name: string;
  description?: string;
  enabled: boolean;
  priority: number;
  conditions: RuleCondition[];
  actions: RuleAction[];
  stopOnMatch?: boolean;
}

// ─── Email shape (subset of IrisEmailDoc that the engine inspects) ──

export interface RuleEmailInput {
  id: string;
  raw: {
    sender: string;
    subject: string;
    body_text?: string;
    has_attachments?: boolean;
  };
  classification: {
    category: ClassificationType;
    sentiment?: SentimentLevel;
    suggestedAction?: SuggestedAction;
  };
  attachments?: Array<{
    detectedType?: string;
    amount?: string;
    extractedText?: string;
  }>;
  score?: number;
  /** Regole già applicate a questa email (per idempotenza) */
  appliedRules?: string[];
}

// ─── Evaluation result ──────────────────────────────────────────────

export interface EvaluationResult {
  matchedRule: IrisRule | null;
  actions: RuleAction[];
  /** se più regole matchano e stopOnMatch=false, qui ce ne sono altre */
  additionalMatches?: IrisRule[];
  /** dati estratti da extract_data condition (se presenti) */
  extractedData?: Record<string, string>;
}

// ─── Execution result ───────────────────────────────────────────────

export interface ActionExecutionResult {
  action: RuleAction;
  ok: boolean;
  detail?: string;
  error?: string;
}

export interface ExecutionResult {
  emailId: string;
  ruleId: string;
  results: ActionExecutionResult[];
  ok: boolean;
}

// ─── Action Runner (dependency-injected) ────────────────────────────

export interface ActionRunner {
  writeLavagna(args: {
    to: string;
    type: string;
    payload: Record<string, unknown>;
    priority: LavagnaPriority;
    sourceEmailId: string;
  }): Promise<{ id: string }>;

  notifyEcho(args: {
    channel: string;
    text: string;
    sourceEmailId: string;
  }): Promise<{ id: string }>;

  archiveEmail(args: { emailId: string }): Promise<void>;

  tagEmail(args: { emailId: string; tags: string[] }): Promise<void>;

  setEmailPriority(args: {
    emailId: string;
    priority: LavagnaPriority;
  }): Promise<void>;

  markRuleApplied(args: {
    emailId: string;
    ruleId: string;
  }): Promise<void>;
}

// ─── Rules loader ───────────────────────────────────────────────────

export interface RulesSource {
  loadEnabled(): Promise<IrisRule[]>;
}

// ─── Engine ─────────────────────────────────────────────────────────

export class RuleEngine {
  private rules: IrisRule[] = [];

  constructor(
    private readonly source: RulesSource,
    private readonly runner: ActionRunner,
  ) {}

  /** Loads rules from the source, sorts by priority desc. */
  async loadRules(): Promise<IrisRule[]> {
    const all = await this.source.loadEnabled();
    this.rules = all
      .filter((r) => r.enabled)
      .sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));
    return this.rules;
  }

  /** In-memory rules (for tests / re-evaluation). */
  setRules(rules: IrisRule[]): void {
    this.rules = [...rules]
      .filter((r) => r.enabled)
      .sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));
  }

  getRules(): IrisRule[] {
    return [...this.rules];
  }

  /** Evaluate all rules on a single email. Returns the first match (or null). */
  evaluate(email: RuleEmailInput): EvaluationResult {
    const additional: IrisRule[] = [];
    let firstMatch: IrisRule | null = null;
    let extracted: Record<string, string> | undefined;

    for (const rule of this.rules) {
      if (!matchAllConditions(email, rule.conditions)) continue;
      if (!firstMatch) {
        firstMatch = rule;
        // Extract data from extract_data actions (used by downstream actions)
        extracted = extractDataFromEmail(email, rule.actions);
        if (rule.stopOnMatch !== false) break;
      } else {
        additional.push(rule);
      }
    }

    return {
      matchedRule: firstMatch,
      actions: firstMatch ? firstMatch.actions : [],
      additionalMatches: additional,
      extractedData: extracted,
    };
  }

  /**
   * Execute matched actions for one email.
   * Skips re-execution if email.appliedRules already contains the ruleId
   * (idempotency guard).
   */
  async execute(
    email: RuleEmailInput,
    rule: IrisRule,
    extractedData?: Record<string, string>,
  ): Promise<ExecutionResult> {
    const results: ActionExecutionResult[] = [];

    if (email.appliedRules?.includes(rule.id)) {
      // Already applied to this email: no-op (avoids duplicate Lavagna writes).
      return {
        emailId: email.id,
        ruleId: rule.id,
        results: [{ action: { type: "tag_email" }, ok: true, detail: "already-applied" }],
        ok: true,
      };
    }

    for (const action of rule.actions) {
      try {
        const detail = await this.runAction(email, action, rule, extractedData);
        results.push({ action, ok: true, detail });
      } catch (e) {
        results.push({
          action,
          ok: false,
          error: e instanceof Error ? e.message : String(e),
        });
      }
    }

    // Mark applied even if some sub-actions failed: we don't want to keep
    // re-running successful side-effects on every pipeline run.
    try {
      await this.runner.markRuleApplied({ emailId: email.id, ruleId: rule.id });
    } catch (e) {
      results.push({
        action: { type: "tag_email" },
        ok: false,
        error: `mark applied failed: ${e instanceof Error ? e.message : String(e)}`,
      });
    }

    return {
      emailId: email.id,
      ruleId: rule.id,
      results,
      ok: results.every((r) => r.ok),
    };
  }

  private async runAction(
    email: RuleEmailInput,
    action: RuleAction,
    rule: IrisRule,
    extractedData?: Record<string, string>,
  ): Promise<string> {
    switch (action.type) {
      case "write_lavagna": {
        if (!action.to) throw new Error("write_lavagna requires 'to'");
        const payload = {
          ...(action.payload ?? {}),
          ...(extractedData ?? {}),
          emailId: email.id,
          subject: email.raw.subject,
          sender: email.raw.sender,
          ruleId: rule.id,
          ruleName: rule.name,
        };
        const r = await this.runner.writeLavagna({
          to: action.to,
          type: action.messageType ?? "iris_event",
          payload,
          priority: action.priority ?? "normal",
          sourceEmailId: email.id,
        });
        return `lavagna:${r.id} → ${action.to}`;
      }
      case "notify_echo": {
        const text = action.text ?? defaultNotifyText(email, rule);
        const r = await this.runner.notifyEcho({
          channel: action.channel ?? "wa",
          text,
          sourceEmailId: email.id,
        });
        return `echo:${action.channel ?? "wa"}:${r.id}`;
      }
      case "archive_email": {
        await this.runner.archiveEmail({ emailId: email.id });
        return "archived";
      }
      case "tag_email": {
        await this.runner.tagEmail({ emailId: email.id, tags: action.tags ?? [] });
        return `tagged:${(action.tags ?? []).join(",")}`;
      }
      case "set_priority": {
        await this.runner.setEmailPriority({
          emailId: email.id,
          priority: action.priority ?? "normal",
        });
        return `priority:${action.priority ?? "normal"}`;
      }
      case "extract_data": {
        // Already processed in evaluate() and merged into Lavagna payload.
        // Nothing to do here at execution time.
        return "extracted (in-evaluate)";
      }
      default: {
        // Exhaustiveness: surface unsupported action types loudly.
        const _exhaustive: never = action.type;
        throw new Error(`Unsupported action type: ${_exhaustive}`);
      }
    }
  }
}

// ─── Pure helpers (testable in isolation) ───────────────────────────

function defaultNotifyText(email: RuleEmailInput, rule: IrisRule): string {
  return `[IRIS · ${rule.name}] ${email.raw.subject || "(senza oggetto)"} — ${email.raw.sender}`;
}

export function matchAllConditions(
  email: RuleEmailInput,
  conditions: RuleCondition[],
): boolean {
  if (!conditions || conditions.length === 0) return false;
  return conditions.every((c) => matchOne(email, c));
}

function fieldValue(email: RuleEmailInput, field: ConditionField): unknown {
  switch (field) {
    case "sender": return email.raw.sender ?? "";
    case "subject": return email.raw.subject ?? "";
    case "category": return email.classification.category;
    case "sentiment": return email.classification.sentiment ?? "neutro";
    case "suggestedAction": return email.classification.suggestedAction;
    case "hasAttachments":
      return Boolean(email.raw.has_attachments) || (email.attachments?.length ?? 0) > 0;
    case "attachmentType": {
      // returns array of attachment types
      return (email.attachments ?? []).map((a) => a.detectedType ?? "");
    }
    case "score": return email.score ?? 0;
  }
}

function matchOne(email: RuleEmailInput, c: RuleCondition): boolean {
  const raw = fieldValue(email, c.field);
  const ci = c.caseInsensitive !== false;
  const norm = (v: unknown): string =>
    ci ? String(v ?? "").toLowerCase() : String(v ?? "");

  switch (c.op) {
    case "equals":
      return Array.isArray(raw)
        ? raw.some((x) => norm(x) === norm(c.value))
        : norm(raw) === norm(c.value);

    case "not_equals":
      return Array.isArray(raw)
        ? !raw.some((x) => norm(x) === norm(c.value))
        : norm(raw) !== norm(c.value);

    case "contains":
      if (Array.isArray(raw)) return raw.some((x) => norm(x).includes(norm(c.value)));
      return norm(raw).includes(norm(c.value));

    case "not_contains":
      if (Array.isArray(raw)) return !raw.some((x) => norm(x).includes(norm(c.value)));
      return !norm(raw).includes(norm(c.value));

    case "in": {
      const arr = Array.isArray(c.value) ? c.value.map(String) : [String(c.value ?? "")];
      const arrN = arr.map((v) => (ci ? v.toLowerCase() : v));
      if (Array.isArray(raw)) return raw.some((x) => arrN.includes(norm(x)));
      return arrN.includes(norm(raw));
    }
    case "not_in": {
      const arr = Array.isArray(c.value) ? c.value.map(String) : [String(c.value ?? "")];
      const arrN = arr.map((v) => (ci ? v.toLowerCase() : v));
      if (Array.isArray(raw)) return !raw.some((x) => arrN.includes(norm(x)));
      return !arrN.includes(norm(raw));
    }

    case "matches": {
      try {
        const re = new RegExp(String(c.value ?? ""), ci ? "i" : "");
        if (Array.isArray(raw)) return raw.some((x) => re.test(String(x ?? "")));
        return re.test(String(raw ?? ""));
      } catch {
        return false;
      }
    }

    case "gt":  return Number(raw) >  Number(c.value);
    case "gte": return Number(raw) >= Number(c.value);
    case "lt":  return Number(raw) <  Number(c.value);
    case "lte": return Number(raw) <= Number(c.value);

    case "is_true":  return Boolean(raw) === true;
    case "is_false": return Boolean(raw) === false;
  }
}

/**
 * Run extract_data actions inline (during evaluate) so the extracted
 * fields are already available when write_lavagna runs.
 *
 * Patterns are JS regex strings; first capture group becomes the value.
 * Searches: body_text, subject, attachments.extractedText.
 */
export function extractDataFromEmail(
  email: RuleEmailInput,
  actions: RuleAction[],
): Record<string, string> | undefined {
  const out: Record<string, string> = {};
  let any = false;
  const haystacks: string[] = [
    email.raw.subject ?? "",
    email.raw.body_text ?? "",
    ...(email.attachments ?? []).map((a) => a.extractedText ?? ""),
  ];
  for (const action of actions) {
    if (action.type !== "extract_data") continue;
    const patterns = action.extractPatterns ?? {};
    for (const [key, pat] of Object.entries(patterns)) {
      try {
        const re = new RegExp(pat, "i");
        for (const h of haystacks) {
          const m = re.exec(h);
          if (m) {
            out[key] = m[1] ?? m[0];
            any = true;
            break;
          }
        }
      } catch {
        // Bad regex — skip silently.
      }
    }
  }
  return any ? out : undefined;
}

File: projects/iris/scripts/rule_engine.py
"""
RuleEngine — porting Python di src/rules/RuleEngine.ts.

Stessa semantica:
  - Conditions: AND su `conditions[]`, op support: equals, not_equals,
    contains, not_contains, in, not_in, matches (regex), gt/gte/lt/lte,
    is_true, is_false. case_insensitive default True.
  - Fields: sender, subject, category, sentiment, suggestedAction,
    hasAttachments, attachmentType (lista), score.
  - Rules ordinate per priority desc.
  - Idempotenza: skippa email con ruleId già in `appliedRules`.

Esegue azioni tramite ActionRunner iniettato. Per la pipeline IRIS, il
runner concreto scrive su Firestore (Lavagna `nexo_lavagna`, marca
`appliedRules` su `iris_emails`, opzionalmente flagga archived).

NB: i Colleghi destinatari (ares/charta/dikea/echo) potrebbero non
esistere ancora come servizi attivi. La Lavagna è bus async: i messaggi
restano in `pending` finché un consumer li picka. Coerente con il
design "lazy materialization".
"""
from __future__ import annotations

import re
from typing import Any, Callable, Iterable, Optional


# ─── Conditions ──────────────────────────────────────────────────

VALID_OPS = {
    "equals", "not_equals", "contains", "not_contains",
    "in", "not_in", "matches", "gt", "gte", "lt", "lte",
    "is_true", "is_false",
}

VALID_FIELDS = {
    "sender", "subject", "category", "sentiment",
    "suggestedAction", "hasAttachments", "attachmentType", "score",
}


def _field_value(email: dict, field: str) -> Any:
    raw = email.get("raw") or {}
    cls = email.get("classification") or {}
    if field == "sender":
        return raw.get("sender") or ""
    if field == "subject":
        return raw.get("subject") or ""
    if field == "category":
        return cls.get("category")
    if field == "sentiment":
        return cls.get("sentiment") or "neutro"
    if field == "suggestedAction":
        return cls.get("suggestedAction")
    if field == "hasAttachments":
        return bool(raw.get("has_attachments")) or bool(email.get("attachments"))
    if field == "attachmentType":
        return [(a.get("detectedType") or "") for a in (email.get("attachments") or [])]
    if field == "score":
        return email.get("score") or 0
    return None


def _norm(v: Any, ci: bool) -> str:
    s = "" if v is None else str(v)
    return s.lower() if ci else s


def match_one(email: dict, c: dict) -> bool:
    field = c.get("field")
    op = c.get("op")
    if field not in VALID_FIELDS or op not in VALID_OPS:
        return False
    val = c.get("value")
    ci = c.get("caseInsensitive", True)
    raw = _field_value(email, field)

    if op in ("equals", "not_equals"):
        target = _norm(val, ci)
        if isinstance(raw, list):
            hit = any(_norm(x, ci) == target for x in raw)
        else:
            hit = _norm(raw, ci) == target
        return hit if op == "equals" else not hit

    if op in ("contains", "not_contains"):
        target = _norm(val, ci)
        if isinstance(raw, list):
            hit = any(target in _norm(x, ci) for x in raw)
        else:
            hit = target in _norm(raw, ci)
        return hit if op == "contains" else not hit

    if op in ("in", "not_in"):
        seq = val if isinstance(val, (list, tuple)) else [val]
        seq_norm = [_norm(x, ci) for x in seq]
        if isinstance(raw, list):
            hit = any(_norm(x, ci) in seq_norm for x in raw)
        else:
            hit = _norm(raw, ci) in seq_norm
        return hit if op == "in" else not hit

    if op == "matches":
        try:
            re_obj = re.compile(str(val or ""), re.IGNORECASE if ci else 0)
        except re.error:
            return False
        if isinstance(raw, list):
            return any(bool(re_obj.search(str(x or ""))) for x in raw)
        return bool(re_obj.search(str(raw or "")))

    if op in ("gt", "gte", "lt", "lte"):
        try:
            a, b = float(raw), float(val)
        except (TypeError, ValueError):
            return False
        return {"gt": a > b, "gte": a >= b, "lt": a < b, "lte": a <= b}[op]

    if op == "is_true":
        return bool(raw) is True
    if op == "is_false":
        return bool(raw) is False

    return False


def match_all_conditions(email: dict, conditions: Iterable[dict]) -> bool:
    cs = list(conditions or [])
    if not cs:
        return False
    return all(match_one(email, c) for c in cs)


# ─── Data extraction (extract_data action, evaluated inline) ────

def extract_data_from_email(email: dict, actions: list[dict]) -> dict:
    out: dict[str, str] = {}
    raw = email.get("raw") or {}
    haystacks = [
        raw.get("subject") or "",
        raw.get("body_text") or "",
    ]
    for a in (email.get("attachments") or []):
        if a.get("extractedText"):
            haystacks.append(a["extractedText"])
    for action in actions:
        if action.get("type") != "extract_data":
            continue
        for key, pat in (action.get("extractPatterns") or {}).items():
            try:
                regex = re.compile(pat, re.IGNORECASE)
            except re.error:
                continue
            for h in haystacks:
                m = regex.search(h)
                if m:
                    out[key] = m.group(1) if m.groups() else m.group(0)
                    break
    return out


# ─── Engine ─────────────────────────────────────────────────────

class RuleEngine:
    """
    runner: oggetto con metodi:
      - write_lavagna(to, message_type, payload, priority, source_email_id) -> str (id)
      - notify_echo(channel, text, source_email_id) -> str (id)
      - archive_email(email_id) -> None
      - tag_email(email_id, tags) -> None
      - set_priority(email_id, priority) -> None
      - mark_rule_applied(email_id, rule_id) -> None
    """

    def __init__(self, runner: Any, rules_loader: Optional[Callable[[], list[dict]]] = None):
        self.runner = runner
        self.rules_loader = rules_loader
        self.rules: list[dict] = []

    def load_rules(self) -> list[dict]:
        all_rules = self.rules_loader() if self.rules_loader else []
        self.rules = sorted(
            [r for r in all_rules if r.get("enabled")],
            key=lambda r: int(r.get("priority", 0)),
            reverse=True,
        )
        return self.rules

    def set_rules(self, rules: list[dict]) -> None:
        self.rules = sorted(
            [dict(r) for r in rules if r.get("enabled")],
            key=lambda r: int(r.get("priority", 0)),
            reverse=True,
        )

    def evaluate(self, email: dict) -> dict:
        first = None
        extras: list[dict] = []
        extracted: dict = {}
        for rule in self.rules:
            if not match_all_conditions(email, rule.get("conditions") or []):
                continue
            if first is None:
                first = rule
                extracted = extract_data_from_email(email, rule.get("actions") or [])
                if rule.get("stopOnMatch", True):
                    break
            else:
                extras.append(rule)
        return {
            "matchedRule": first,
            "actions": (first or {}).get("actions") or [],
            "additionalMatches": extras,
            "extractedData": extracted or None,
        }

    def execute(self, email: dict, rule: dict, extracted: Optional[dict] = None) -> dict:
        results = []
        rule_id = rule.get("id") or "?"
        if rule_id in (email.get("appliedRules") or []):
            return {
                "emailId": email.get("id"),
                "ruleId": rule_id,
                "results": [{"action": {"type": "tag_email"}, "ok": True, "detail": "already-applied"}],
                "ok": True,
            }

        for action in rule.get("actions") or []:
            try:
                detail = self._run_action(email, action, rule, extracted)
                results.append({"action": action, "ok": True, "detail": detail})
            except Exception as e:
                results.append({"action": action, "ok": False, "error": str(e)})

        try:
            self.runner.mark_rule_applied(email.get("id"), rule_id)
        except Exception as e:
            results.append({"action": {"type": "tag_email"}, "ok": False,
                            "error": f"mark applied failed: {e}"})

        return {
            "emailId": email.get("id"),
            "ruleId": rule_id,
            "results": results,
            "ok": all(r["ok"] for r in results),
        }

    def _run_action(self, email: dict, action: dict, rule: dict, extracted: Optional[dict]) -> str:
        atype = action.get("type")
        if atype == "write_lavagna":
            to = action.get("to")
            if not to:
                raise ValueError("write_lavagna requires 'to'")
            payload = {
                **(action.get("payload") or {}),
                **(extracted or {}),
                "emailId": email.get("id"),
                "subject": (email.get("raw") or {}).get("subject"),
                "sender": (email.get("raw") or {}).get("sender"),
                "ruleId": rule.get("id"),
                "ruleName": rule.get("name"),
            }
            mid = self.runner.write_lavagna(
                to=to,
                message_type=action.get("messageType") or "iris_event",
                payload=payload,
                priority=action.get("priority") or "normal",
                source_email_id=email.get("id"),
            )
            return f"lavagna:{mid} → {to}"
        if atype == "notify_echo":
            text = action.get("text") or _default_notify_text(email, rule)
            mid = self.runner.notify_echo(
                channel=action.get("channel") or "wa",
                text=text,
                source_email_id=email.get("id"),
            )
            return f"echo:{action.get('channel') or 'wa'}:{mid}"
        if atype == "archive_email":
            self.runner.archive_email(email.get("id"))
            return "archived"
        if atype == "tag_email":
            tags = action.get("tags") or []
            self.runner.tag_email(email.get("id"), tags)
            return f"tagged:{','.join(tags)}"
        if atype == "set_priority":
            prio = action.get("priority") or "normal"
            self.runner.set_priority(email.get("id"), prio)
            return f"priority:{prio}"
        if atype == "extract_data":
            return "extracted (in-evaluate)"
        raise ValueError(f"Unsupported action type: {atype}")


def _default_notify_text(email: dict, rule: dict) -> str:
    raw = email.get("raw") or {}
    return f"[IRIS · {rule.get('name')}] {raw.get('subject') or '(senza oggetto)'} — {raw.get('sender')}"
